File: Matlab/TrainNeuralNet.m
function [ validationWeights, weights, validationErrors] = TrainNeuralNet( trainData, trainLabels, validationData, validationLabels, nil, maxNetworkIterations, theta, improveTheta)
%TRAINNEURALNET trains a feed-forward artificial neural network using
%backpropagation learning. The network automatically stops training when
%performance decreases on the validation set.
%INPUT:
%trainData: 3D matrix containing greyscale training images
%trainLabels: 2D matrix containing one, binary label per row
%validationData: a separate subset of data used to predict network accuracy
%validationLabels: 2D matrix containing one, binary label per row
%nil: neurons-in-layer. Vector containing the number of neurons to place in
%each layer.
%maxNetworkIterations: Maximim number of times the network will run before
%stopping, if not stopped by increase in error on the validation set
%earlier.
%theta: 3d parameter matrix. Use nan if training a new network. Use a
%previously obtained value in order to improve upon that value.
%improveTheta: boolean. If theta has a value and improveTheta is 1, the
%input value of theta will be improved upon.
%OUTPUT:
%[validationWeights, weights, validationErrors]

[trainWidth, trainHeight, numTrainingSamples]= size(trainData);
[validationWidth, validationHeight, numValidationSamples]=size(validationData);
validationErrors=[];

L = length(nil); %Number of layers. Note that there are L-1 layers in theta

%Training constant information
alpha = 1;

if isnan(theta) && ~improveTheta
    %Initialize random parameters -.5 to .5
    for n = 1:length(nil)-1
        theta(1:nil(n), 1:nil(n+1), n) = rand(nil(n), nil(n+1)) -0.5; %#ok<AGROW>
    end
end

%Initialize activations
activations(1:length(nil), 1:max(nil)) = 0;

%Initialize activation error matrix (2D)
deltaLC = zeros(size(activations,2), size(activations,1));

%Sigmoid hypothesis
h=@(x) 1./(1+exp(-x));

for iters=1:maxNetworkIterations
    
    %Initialize weight change matrix (3D). Must be done each iteration.
    deltaUC = zeros(size(theta,1), size(theta,2), size(theta,3));
    
    %Reset information
    error=0;
    display(iters); %Should be left in actual run code in order to display progress. This can be SLOW (days or weeks)
    for m=1:numTrainingSamples
        %Forward Propagation. This step is a factor of 1-3 faster than backprop.
        activations(1, 1:nil(1)) = reshape(trainData(:, :, m), trainWidth.*trainHeight, 1); %Unroll data;
        for layer = 2:L
            activations(layer, 1:nil(layer)) =  h( activations(layer-1, 1:nil(layer-1))*theta(1:nil(layer-1), 1:nil(layer), layer-1) );
        end
        
        %Backpropagation: Find theta updates
        %Output layer is easy, but must be done separately
        outputLCVec=activations(L, 1:nil(L))' - (trainLabels(m, 1:nil(L)))';
        deltaLC(1:nil(L), L) = outputLCVec;
        deltaUC(1:nil(L-1), 1:nil(L), L-1) = deltaUC(1:nil(L-1), 1:nil(L), L-1) + activations(L-1, 1:nil(L-1))'*deltaLC(1:nil(L),L)';
        
        %Harvest squared error in final layer
        error=error+sum(outputLCVec.^2);
        
        %Hidden layers
        for layer=L-1:-1:2 %No error in the input layer
            activationsInPreviousLayer=activations(layer, 1:nil(layer));
            
            %Application of analytical partial error derivative with respect to the weights.
            deltaLC(1:nil(layer),layer) = (theta(1:nil(layer), 1:nil(layer+1), layer)*deltaLC(1:nil(layer+1),layer+1))'.*activationsInPreviousLayer.*(1-activationsInPreviousLayer);
            
            %3D matrix subtraction to update all parameters
            deltaUC(1:nil(layer-1), 1:nil(layer), layer-1) = deltaUC(1:nil(layer-1), 1:nil(layer), layer-1) + activations(layer-1, 1:nil(layer-1))'*deltaLC(1:nil(layer),layer)';
        end
    end
    
    
    for layer=1:L-1
        tempDeltaUCLayer = deltaUC(1:nil(layer), 1:nil(layer+1), layer);
        deltaUC(1:nil(layer), 1:nil(layer+1), layer) = tempDeltaUCLayer./max(max(abs(tempDeltaUCLayer)));
    end
    
    
    %Update to theta. Performed in a manner such that error always goes
    %down (error minimum is never overshot).
    done=0;
    while ~done
        newTheta = theta - alpha.*deltaUC; %Theta update
        newError=0;
        for m=1:numTrainingSamples %Calculate new error
            %Forward Propagation. No need to store activation matrix.
            tempActivationVec = reshape(trainData(:, :, m), 1, trainHeight.*trainWidth); %Unroll data;
            for layer = 2:L
                tempActivationVec =  h(tempActivationVec*newTheta(1:nil(layer-1), 1:nil(layer), layer-1));
            end
            
            %Backpropagation for first layer only (find errors)
            newError=newError + sum((tempActivationVec - (trainLabels(m, :))).^2); %check order;
        end
        
        %Implementation of momentum for more efficient learning
        if newError<error
            theta=newTheta;
            alpha=alpha.*2;
            error=newError;
            done=1;
            display(alpha);
        else %Error has increased
            alpha=alpha./2;
        end
    end
    
    %Implementation of validation for early stopping to avoid overfitting
    validationErr=0;
    for m=1:numValidationSamples %Calculate new error
        %Forward Propagation. No need to store activation matrix.
        tempActivationVec = reshape(validationData(:, :, m), 1, validationHeight.*validationWidth); %Unroll data;
        for layer = 2:L
            tempActivationVec =  h(tempActivationVec*theta(1:nil(layer-1), 1:nil(layer), layer-1));
        end
        
        %Backpropagation for first layer only (find errors)
        validationErr=validationErr + sum((tempActivationVec - (validationLabels(m, :))).^2);
    end
    
    %Validation errors are useful for plotting purposes as well as for
    %stopping and should be saved.
    
    %Not preallocating here does little harm and helps later syntax
    validationErrors=[validationErrors, validationErr];
    
    %Choose the set of weights (theta) that better generalizes
    if length(validationErrors)>1 && validationErrors(end)<validationErrors(end-1)
        validationWeights=theta;
    end
    
    if length(validationErrors)>15 && max(validationErrors(end-15:end))==validationErrors(end)
        %Has been overfitting. Stop to save time.
        break;
    end
    
end

weights=newTheta;

end

File: Python/TrainNeuralNet.py
import numpy as np
import NeuralNetSupportFunctions as ANNSupport

#Handles the low-level FFANN training process
def RunTrainNeuralNet(trainData, trainLabels, validationData, validationLabels, nil, maxNetworkIterations, theta, improveTheta):
    trainShape=trainData.shape
    trainSamples=trainShape[0]
    trainPixels=trainShape[1]

    validationShape=validationData.shape
    validationSamples=validationShape[0]
    validationPixels=validationShape[1]
    validationErrors=[]
    validationWeights=0
    validationLookback=2
    validationSampleRate=20

    L=nil.size #Number of layers. Note that there are L-1 layers in theta
    alpha=1 #Training constant information
    alphaTol=0.00000001

    if theta==0 and (not improveTheta):
        theta=np.zeros((np.max(nil[0:-1]), np.max(nil[1:]), L-1))
        for n in np.arange(L-1):
            #Initialize weight layers between -0.5 and 0.5
            theta[0:nil[n], 0:nil[n+1], n]=np.random.rand(nil[n], nil[n+1])-0.5

    #Initialize activation error matrix (2D).
    #Note that entries are stored row-wise in activations
    #and column-wise in deltaLC
    deltaLC=np.zeros((np.max(nil),L))

    #Begin main loop
    error=999999999
    iters=-1
    while iters<maxNetworkIterations and error>0.00001 and alpha>alphaTol:
        iters=iters+1;
        #Initialize 3D weight update matrix. Must be done each iteration
        deltaUC=np.zeros(theta.shape)

        #Reset additional iteration specific variables
        error=0
        print("Iterations: "+str(iters)) #Useful feedback for long training sessions

        for m in np.arange(trainSamples):
            #Forward propagation
            activations = ANNSupport.forwardProp(trainData[m, 0:trainPixels], trainPixels, nil, L, theta)

            #Backpropagation
            #Output layer is easy but must be done separately
            errorVec, squaredError=ANNSupport.checkPrediction(activations[-1, 0:nil[-1]], trainLabels[m], nil, 0)
            deltaLC[0:nil[L-1],L-1]=errorVec.T
            deltaUC[0:nil[-2],0:nil[-1],-1] += np.outer(activations[-2, 0:nil[-2]], deltaLC[0:nil[-1],-1])
            #Harvest squared error in final layer
            error += squaredError
            #Hidden layers. No update to input layer (no error)
            for layer in np.arange(L-2, 0, -1):
                activationsInPrevious = activations[layer, 0:nil[layer]]
                deltaLC[0:nil[layer], layer] = np.dot(theta[0:nil[layer],0:nil[layer+1],layer],deltaLC[0:nil[layer+1],layer+1]).T * activationsInPrevious * (1-activationsInPrevious)
                deltaUC[0:nil[layer-1],0:nil[layer],layer-1] += np.outer(activations[layer-1, 0:nil[layer-1]], deltaLC[0:nil[layer],layer])

        #Scale weights by layer
        for layer in np.arange(L-1):
            tempDeltaUCLayer = deltaUC[0:nil[layer],0:nil[layer+1], layer]
            deltaUC[0:nil[layer],0:nil[layer+1],layer] = tempDeltaUCLayer/np.max(np.abs(tempDeltaUCLayer))

        #Update theta. Error always decreases in batch mode
        done=0
        while not done:
            newTheta = theta - alpha*deltaUC
            newError=0
            for m in np.arange(trainSamples):
                #Forward propagation
                activations = ANNSupport.forwardProp(trainData[m, 0:trainPixels], trainPixels, nil, L, newTheta)
                #Backpropagation for final layer to calculate error
                errorVec, squaredError=ANNSupport.checkPrediction(activations[-1, 0:nil[-1]], trainLabels[m], nil, 0)
                newError += squaredError

            #Implementation of momentum
            if newError<error:
                theta=newTheta;
                alpha*=2;
                error=newError;
                done=1;
                print('new error = '+str(error))
                print('alpha = '+str(alpha));
            else: #Error has increased
                alpha/=2

        #Evaluation of error on validation set
        if validationSamples>0 and iters%validationSampleRate==0:
            validationErr=0
            for m in np.arange(validationSamples):
                #Forward propagation
                activations = ANNSupport.forwardProp(validationData[m, 0:validationPixels], validationPixels, nil, L, theta)
                #Backpropagation for final layer to calculate error
                squaredError=ANNSupport.checkPrediction(activations[-1, 0:nil[-1]], validationLabels[m], nil, 1)
                validationErr+=squaredError

            #Validation errors should be saved for early stopping and plotting
            validationErrors.append(validationErr)

            #Save the set of weights that performs better on the validations set
            if len(validationErrors)>1 and validationErrors[-1]<validationErrors[-2]:
                    validationWeights=theta

            #Check for overfitting
            if len(validationErrors)>validationLookback and np.max(validationErrors[-validationLookback:])==validationErrors[-1]:
                #Stop execution
                print("Overfitting Detected. Returning Weights")
                break

    #End of outer iteration loop
    if iters==maxNetworkIterations:
        print('Max Iterations Reached. Returning Weights')
    elif error<=0.00001:
        print('Error Tolerance Reached')
    elif alpha<=alphaTol:
        print('Learning Rate Tolerance Reached')

    if np.sum(validationWeights)==0:
        print('Training Weights Returned')
        return theta, validationErrors
    else:
        print('Validation Weights Returned')
        return validationWeights, validationErrors



